src/sympc/protocol/falcon/falcon.py

```diff
@@ -404,3 +431,289 @@ def multiplication_protocol(
+        session = x[0].session
+        ptr_list: List = []
+        for session_ptr in session.session_ptrs:
+            sh_ptr = session_ptr.prrs_generate_random_share(
+                shape=shape, ring_size=str(2)
+            )
+            ptr_list.append(sh_ptr)
+
+        beta_2 = MPCTensor(
+            shares=ptr_list, session=session, shape=shape
+        )  # shares of random bit
+        beta_p = ABY3.bit_injection(
+            beta_2, session, PRIME_NUMBER
+        )  # shares of random bit in Zp.
+        m = Falcon._random_prime_group(session, shape)
+        if isinstance(r, np.ndarray):
+            beta_2.to_numpy("bool_")
+            beta_p.to_numpy("uint8")
+            m.to_numpy("uint8")
+
+        u = [0 for i in range(len(x))]
+        w = [0 for i in range(len(x))]
+        c = [0 for i in range(len(x))]
+        for i in range(len(x) - 1, -1, -1):
+            r_i = r >> i & 1  # bit at ith position
+            u[i] = (1 - 2 * beta_p) * (x[i] - r_i)
+            w[i] = x[i] ^ r_i
+            c[i] = u[i] + 1 + sum(w[i + 1 :])
+
+        d = m * (math.prod(c))
+
+        d_val = d.reconstruct(decode=False)  # plaintext d.
+        d_val[d_val != 0] = 1  # making all non zero values as 1.
+
+        beta_prime = d_val
+
+        return beta_2 + beta_prime
+
+    @staticmethod
+    def wrap2(a: np.ndarray, b: np.ndarray) -> np.ndarray:
+        """Computes wrap on the input numpy array.
+
+        Args:
+            a (np.ndarray): Input numpy array
+            b (np.ndarray): Input numpy array.
+
+        Returns:
+            result (np.ndarray): Boolean array ,True if there is a carry.
+
+        Raises:
+            ValueError: If the input values is not a numpy array.
+            ValueError: If signed numpy array are provided as input.
+
+        wrap2 calucaties the carry bit on addition of two values a+b.
+        """
+        if not isinstance(a, np.ndarray) or not isinstance(b, np.ndarray):
+            raise ValueError("Input value must be a numpy array for wrap2.")
+
+        if a.dtype.kind == "i" or b.dtype.kind == "i":
+            raise ValueError("Wrap2 works only for signed numbers.")
+
+        max_val = np.iinfo(a.dtype).max
+
+        return a > max_val - b
+
+    @staticmethod
+    def wrap3(a: np.ndarray, b: np.ndarray, c: np.ndarray) -> np.ndarray:
+        """Computes wrap on the input numpy array.
+
+        Args:
+            a (np.ndarray): Input numpy array
+            b (np.ndarray): Input numpy array.
+            c (np.ndarray): Input numpy array.
+
+        Returns:
+            result (np.ndarray): Modulo reduction of the exact wrap function.
+
+        Raises:
+            ValueError: If the input values is not a numpy array.
+            ValueError: If signed numpy array are provided as input.
+
+        wrap3 calucaties the carry bit on addition of three values a+b+c(mod 2).
+        """
+        if not isinstance(a, np.ndarray) or not isinstance(b, np.ndarray):
```

Comment:
Q: shouldn't we check for c too?

Reply:
Yeah we should for c,I would change it :+1:

Reply:
Done